### PR TITLE
Material: Add missing blend factors for constant color and constant alpha

### PR DESCRIPTION
**Pull request by christjt**
Related issue: #23741 

**Description**
Adds missing blend factors (`ConstantColor`, `OneMinusConstantColor`, `ConstantAlpha`, `OneMinusConstantAlpha`) and adds a complimenting `blendColor` field to `Material`.

Also updated example and documentation.

This contribution is funded by [Cognite](https://cognite.com).

### CODE REVIEW

**Comment by christjt** [comment]
I see that the example I updated is failing, which is expected since, well... I changed it. Any way for me to update the screenshot, or should I simply discard any changes made to that example?

**Comment by christjt** [comment]
I have updated the test image and e2e-tests passes locally. Anything I need to update?

**Review by gkjohnson — COMMENTED** [review]
Can you elaborate on some use cases for how you plan to use these blend modes? With them just operating with constant values they don't seem that useful at first blush and like they could be implemented other ways. And the example update doesn't show any compelling use cases.

Live link to the updated example:
https://raw.githack.com/christjt/three.js/blend-constants/examples/webgl_materials_blending_custom.html

**Comment by christjt** [comment]
Hey, the main use case I have for this is the ability to blend by a specific alpha while preserving the alpha of the source.

Basically I can set 
- blendDst = ConstantAlpha
- blendDstAlpha = ZeroFactor
- blendSrc = OneMinusConstantAlphaFactor
- blendSrcAlpha = OneFactor

Then let blendColor be (0.0, 0.0, 0.0, 0.5) f.ex.

This is extremely useful because you can use the alpha channel to store other data than transparency. In this case I can blend the colors by the constant alpha, but the actual alpha value of the source will be preserved, which is necessary if you encode some specific byte data into the alpha channel that you want to use at a later stage in your rendering pipeline.

Does this use-case make sense? I'll fix the example (was working fine for me locally, but seems I messed something up as it clearly isn't working as intended in the link you shared). Also I will fix a proper default value of (0, 0, 0, 0).

File: src/constants.js [hunk 1]
```diff
@@ -36,6 +36,10 @@ export const OneMinusDstAlphaFactor = 207;
 export const DstColorFactor = 208;
 export const OneMinusDstColorFactor = 209;
 export const SrcAlphaSaturateFactor = 210;
+export const ConstantColor = 211;
+export const OneMinusConstantColor = 212;
+export const ConstantAlpha = 213;
+export const OneMinusConstantAlpha = 214;
```
Comment (mrdoob) [inline src/constants.js:42]: All the other constants end with `Factor`.
How come the new ones do not?

**Comment by LeviPesin** [comment]
There also are WebGPU's [BlendColorFactor and OneMinusBlendColorFactor](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/renderers/webgpu/constants.js). They should also possibly be added.

**Comment by WestLangley** [comment]
>Does this use-case make sense?
>...but seems I messed something up as it clearly isn't working as intended in the link you shared

Your math, as you described it, does not seem to me to be consistent with your objective.

**Comment by christjt** [comment]
> > Does this use-case make sense?
> > ...but seems I messed something up as it clearly isn't working as intended in the link you shared
> 
> Your math, as you described it, does not seem to me to be consistent with your objective.

I am probably explaining this very poorly. This feature enables users to blend two different objects by a color and/or alpha that is not defined by the objects themselves. 

I'll try to contrive a better example:

Imagine you have some player object that is walking either in-front or behind some wall. When the player walks in-front of the wall you want everything to be rendered opaque, but when the player is behind the wall you want a slight blend such that the player kan be slightly seen through the wall (like some sort of ghost effect or something).

To achieve this effect without drawing the player in two passes. You can draw the player first (writing to the depth buffer), and then draw the wall (with depth test) while blending it by some constant color that you wish for your effect (maybe some red color or something).

You see this type of visualization often in shooter games where they have some visualization where you can see players behind walls (often getting some color tint applied).

Here is an example where you can see the players behind a wall, but tinted red only when behind.
image

**Comment by gkjohnson** [comment]
I'm not sure I'm following the flow of logic for the first person shooter use-case. Do you have a write up on the technique?

But that aside @mrdoob what do you think of adding these? I have a use case for three-gpu-pathtracer that would benefit from the use of a constant alpha factor when averaging new samples into the frame. Right now the output alpha is used for this but I would like to use output alpha to add support for background transparency which this feature would afford.

Here's what I propose:
- Add `Factor` at the end of the enum names so they are consistent
- Add `Material.blendColor` or `Material.constantColor` (just an alternative name)
- Add `Material.blendAlpha` or `Material.constantAlpha`

@mrdoob @Mugen87 what do you think?

File: examples/webgl_materials_blending_custom.html [hunk 2]
```diff
@@ -122,10 +128,12 @@
 						material.blendSrc = blendSrc.constant;
 						material.blendDst = blendDst.constant;
 						material.blendEquation = THREE.AddEquation;
+						material.blendColor = new THREE.Color( params.blendColor );
+						material.blendColor.a = params.alpha;
```
Comment (gkjohnson) [inline examples/webgl_materials_blending_custom.html:132]: There is no `.a` property on `THREE.Color` so it's not valid to add it here.

**Comment by Mugen87** [comment]
@gkjohnson I'm fine with your suggestion. I vote for `blendColor` and `blendAlpha`. To keep the change simple, I would not modify any of the official examples though. 

It's probably best if you create a new PR in favor of this one.

**Comment by gkjohnson** [comment]
Awesome! @christjt are you still interested in getting everything updated for this feature?

**Comment by WestLangley** [comment]
@gkjohnson Please be advised of what @WestLangley [said](https://github.com/mrdoob/three.js/pull/23754#issuecomment-1079954222). I'd welcome your second-opinion.

**Comment by christjt** [comment]
@WestLangley thanks, yeah I messed up some of the potensial solution to the use  case, should just be blending dst and src color by constant alpha and one minus constant alpha and then separately blend the alpha values.

**Comment by gkjohnson** [comment]
@christjt if you have references or sample implementations for your FPS example I'd be interested to see them but the explanation does not line up with what I'm seeing / understanding from the image. But I think these are useful constants either way.

> @gkjohnson Please be advised of what @WestLangley https://github.com/mrdoob/three.js/pull/23754#issuecomment-1079954222. I'd welcome your second-opinion.

Yes I agree that the current demo does not look correct but assuming that is fixed this is a good addition, I think.

**Pull request closed without merging**